client/starwhale/api/_impl/dataset/builder.py

```diff
@@ -482,26 +495,152 @@ def _do_iter_item(self: t.Any) -> t.Generator:
+    use_swds_bin: bool, iter_func: t.Callable
+) -> t.Type[BaseBuildExecutor]:
+    attrs = {"iter_item": iter_func}
+    if use_swds_bin:
         _cls = type(
-            "GenericUserRawHandler",
-            (UserRawBuildExecutor,),
+            "GenericSWDSBinHandler",
+            (SWDSBinBuildExecutor,),
             attrs,
         )
     else:
         _cls = type(
-            "GenericSWDSBinHandler",
-            (SWDSBinBuildExecutor,),
+            "GenericUserRawHandler",
+            (UserRawBuildExecutor,),
             attrs,
         )
-
     return _cls
+
+
+class RowWriter(threading.Thread):
+    def __init__(
+        self,
+        dataset_name: str,
+        dataset_version: str,
+        project_name: str = DEFAULT_PROJECT,
+        workdir: Path = Path(".dataset_tmp"),
+        alignment_bytes_size: int = D_ALIGNMENT_SIZE,
+        volume_bytes_size: int = D_FILE_VOLUME_SIZE,
+        append: bool = False,
+        append_from_version: str = "",
+        append_from_uri: t.Optional[URI] = None,
+        append_with_swds_bin: bool = True,
+    ) -> None:
+        super().__init__(
+            name=f"RowWriter-{dataset_name}-{dataset_version}-{project_name}"
+        )
+
+        self._kw = {
+            "dataset_name": dataset_name,
+            "dataset_version": dataset_version,
+            "project_name": project_name,
+            "workdir": workdir,
+            "alignment_bytes_size": alignment_bytes_size,
+            "volume_bytes_size": volume_bytes_size,
+            "append": append,
+            "append_from_version": append_from_version,
+            "append_from_uri": append_from_uri,
+        }
+
+        self._queue: queue.Queue[t.Optional[DataRow]] = queue.Queue()
+        self._summary = DatasetSummary()
+        self._lock = threading.Lock()
+
+        self._run_exception: t.Optional[Exception] = None
+
+        self.setDaemon(True)
+        self._builder: t.Optional[BaseBuildExecutor] = None
+        if append and append_from_version:
+            _cls = create_generic_cls_by_mode(append_with_swds_bin, self.__iter__)
+            self._builder = _cls(**self._kw)  # type: ignore
+            self.start()
+
+    def _raise_run_exception(self) -> None:
+        if self._run_exception is not None:
+            _e = self._run_exception
+            self._run_exception = None
+            raise threading.ThreadError(f"RowWriter Thread raise exception: {_e}")
+
+    @property
+    def summary(self) -> DatasetSummary:
+        return self._summary
+
+    def __enter__(self) -> RowWriter:
+        return self
+
+    def __exit__(
+        self,
+        type: t.Optional[t.Type[BaseException]],
+        value: t.Optional[BaseException],
+        trace: TracebackType,
+    ) -> None:
+        if value:  # pragma: no cover
+            logger.warning(f"type:{type}, exception:{value}, traceback:{trace}")
+
+        self.close()
+
+    def flush(self) -> None:
+        while not self._queue.empty():
```

Comment:
1. python queue is [thread-safe](https://docs.python.org/3/library/queue.html). thread1 and thread2 use the same queue.
2. RowWriter `flush()` function is designed to block here to ensure the previous `__setitem__` and `__delitem__` instructions are consumed in the current thread context. The flush running thread does not affect other threading run routines, but if others threads are writing a lot of records, the flush waiting time may be longer.
3. RowWriter flush() function is the same as io flush() function, it can be called many times.
4. RowWrtier flush() function is not the close() function. I have already implemented the [close()](https://github.com/star-whale/starwhale/pull/1554/files#diff-acca3ef3c1d4d6d3b4faba346fb1d6f3a128c89ac79d051d0ec563a3db0f360dR605) function, which will call RowWriter thread join() and ensure all records will be consumed.

Reply:
In the previous example scenario:

t1:
```
rw.update(row1)
time.sleep(1)
rw..update(row2)
```

t2:
```
rw.update(row3)
rw.update(row4)
rw.flush()
rw.update(row5)
```

flush() in the t2 thread only ensure the row3 and row4 that are consumed. It will block before updating row5.